### DISCUSSION TITLE
Serenity with Always Encrypted Sql Server

### DISCUSSION THREAD

**Discussion by teafmartins** [serenity-is/Serenity#6365] [Help] [answered]
Hi,

Anyone using Always Encrypted Sql server with Serenity?
Any trouble retrieving results in endpoints? I'm having timeout retrieving encrypted columns.

Thanks

**Comment by edwardch** [comment]
Hey @teafmartins,

I would be happy experiment for you if you could share a simplified explanation of your SQL Server set up?

I'd like to encrypt some columns too, so this may benefit both of us.

Cheers,
Edward

**Comment by teafmartins** [comment]
Hi @edwardch 

I used this guideline:
https://www.toptal.com/sql-server/sql-server-2016-always-encrypted
I just created a key in SSMS and after that the master key column.
![image](https://user-images.githubusercontent.com/17210232/164399807-c2cfa8e2-87d0-46a6-8b6d-ea3c99e909c6.png)
after that I just created the encrypted the column needed:
![image](https://user-images.githubusercontent.com/17210232/164400682-d8fa996c-aaa7-4064-8f11-245639ae9324.png)

When running Serenity everything works find if I don't retrieve encrypted columns in endpoint. If I include this columns I get time out error.

Can you try your side if it works?

**Reply by edwardch** [reply]
Looks interesting, thanks. Let me give it a shot next week. Hopefully get something promising.

Out of interest, do you host production on self managed servers, or are you using managed like Azure SQL or something?

**Comment by teafmartins** [comment, marked as answer]
Always Encrypt works very well reading data from DB but not inserting or updating data.
Inserted and updated must be done using parameterized query. 

How can I do it using serenity? any easy way modifying endpoint?

**Reply by Julian1993ARG** [reply]
Hi, did you find any way to do it?

**Reply by teafmartins** [reply]
Hi Julian, 

Yes, as I said above, you have to change the "MySaveHandler" class in repository and manually perform the update with parameters.
For example, something like:

                string ConnectionString = ConfigurationManager.ConnectionStrings["default"].ConnectionString;
                using (SqlConnection connection = new SqlConnection(ConnectionString))
                {
                    using (var cmd = new SqlCommand("UPDATE [dbo].[Table] SET FieldA=@FieldA WHERE Id=" + Row.Id, connection))
                    {
                        connection.Open();
                        cmd.CommandTimeout = 30;
                        var paramFieldA = new SqlParameter("@FieldA", SqlDbType.NVarChar, 50);
                        paramFieldA.Value = Row.FieldA;
                        cmd.Parameters.Add(paramFieldA);
                        cmd.ExecuteNonQuery();
                    }
                }

I hope this helps.
